### PR TITLE
Feature/remove public key after destroyed

### PR DESCRIPTION
**Pull request by 9547**
### What problem does this PR solve? 
Try to implement #898 

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

Release notes:

```release-note
Remove SSH public key if the cluster was destroyed 
```

### CODE REVIEW

File: components/cluster/command/prune.go [hunk 1]
```diff
@@ -73,7 +73,7 @@ func destroyTombstoneIfNeed(clusterName string, metadata *spec.ClusterMeta, opt
 		return perrs.AddStack(err)
 	}
 
-	nodes, err := operator.DestroyTombstone(ctx, topo, true /* returnNodesOnly */, opt, tlsCfg)
+	nodes, err := operator.DestroyTombstone(ctx, topo, true /* returnNodesOnly */, opt, tlsCfg, ctx.PublicKeyPath)
```
Comment (AstroProfundis) [inline components/cluster/command/prune.go:76]: `ctx` is passed to `DestroyTombstone()` so I don't think it's necessary to pass `ctx.PublicKeyPath` again.

File: components/dm/command/scale_in.go [hunk 2]
```diff
@@ -46,7 +46,7 @@ func newScaleInCmd() *cobra.Command {
 				b.Func(
 					fmt.Sprintf("ScaleInCluster: options=%+v", gOpt),
 					func(ctx *task.Context) error {
-						return ScaleInDMCluster(ctx, metadata.Topology, gOpt)
+						return ScaleInDMCluster(ctx, metadata.Topology, gOpt, ctx.PublicKeyPath)
```
Comment (AstroProfundis) [inline components/dm/command/scale_in.go:49]: ditto

File: pkg/cluster/manager.go [hunk 3]
```diff
@@ -252,7 +252,8 @@ func (m *Manager) RestartCluster(clusterName string, options operator.Options) e
 func (m *Manager) ListCluster() error {
 	names, err := m.specManager.List()
 	if err != nil {
-		return perrs.AddStack(err)
+		log.Warnf("List cluster error %v", err)
+		// return perrs.AddStack(err)
```
Comment (AstroProfundis) [inline pkg/cluster/manager.go:256]: Why not returning here?

File: pkg/cluster/manager.go [hunk 4]
```diff
@@ -381,7 +382,7 @@ func (m *Manager) DestroyCluster(clusterName string, gOpt operator.Options, dest
 			}, tlsCfg)
 		}).
 		Func("DestroyCluster", func(ctx *task.Context) error {
-			return operator.Destroy(ctx, topo, destroyOpt)
+			return operator.Destroy(ctx, topo, ctx.PublicKeyPath, destroyOpt)
```
Comment (AstroProfundis) [inline pkg/cluster/manager.go:385]: ditto, the new param is not necessary.

File: pkg/cluster/operation/destroy.go [hunk 5]
```diff
@@ -171,6 +195,40 @@ func DeleteGlobalDirs(getter ExecutorGetter, host string, options *spec.GlobalOp
...
+	publicKey, err := ioutil.ReadFile(pubKeyPath)
+	if err != nil {
+		return perrs.Trace(err)
+	}
+	pubKey := string(bytes.TrimSpace(publicKey))
+	pubKey = strings.ReplaceAll(pubKey, "/", "\\/")
+	pubKeysFile := executor.FindSSHAuthorizedKeysFile(e)
+
+	c := module.ShellModuleConfig{
+		Command:  fmt.Sprintf("sed -i '/%s/d' %s", pubKey, pubKeysFile),
```
Comment (AstroProfundis) [inline pkg/cluster/operation/destroy.go:211]: I think it might be better to just comment that line.

**Review by AstroProfundis — COMMENTED** [review]
And as we support deploying cluster with user specified SSH key (instead of generating new), so it would be better to check for duplicate during deploy process, and also not deleting the public key by default on scaling in and destroying.

**Comment by lucklove** [comment]
Seems test broken: `cannot stat '~/.tiup/storage/dm/clusters/test_cmd/ssh/id_rsa'`
May related to this PR.

File: components/cluster/command/prune.go (hunk 1 above)
Comment (9547) [reply components/cluster/command/prune.go:76]: The first argument inside `DestoryTombestone()` is `ExecutorGetter` interface: https://github.com/pingcap/tiup/blob/cc41ad5af87bde821ecde48d63dd7c1c40d72ea2/pkg/cluster/operation/destroy.go#L414-L420  So I need to cast it into `task.Context`, but this results in `import cycle` error. Please let me know if I've misssed something.

File: pkg/cluster/manager.go (hunk 3 above)
Comment (9547) [reply pkg/cluster/manager.go:256]: Sorry, this is a mistake, will be reverted back.

File: components/cluster/command/prune.go (hunk 1 above)
Comment (lucklove) [reply components/cluster/command/prune.go:76]: Maybe we can define another interface for the first param of `DestroyTombstone`:
```golang
type SSHExecutorGetter interface {
    ExecutorGetter
    GetSSHKeySet() (string, string) // return both priv and public key
}
```

Comment (9547) [reply components/cluster/command/prune.go:76]: @lucklove Or should we extend the original `ExecutorGetter` interface with:

```go
type ExecutorGetter interface {
	Get(host string) (e executor.Executor)
	GetSSHKeySet() (string, string)
}
```

File: pkg/cluster/operation/destroy.go (hunk 5 above)
Comment (9547) [reply pkg/cluster/operation/destroy.go:211]: Thanks for the advice.

File: components/cluster/command/prune.go (hunk 1 above)
Comment (9547) [reply components/cluster/command/prune.go:76]: @lucklove I did it this way first, if you have any question, please let me know.

Comment (lucklove) [reply components/cluster/command/prune.go:76]: > @lucklove Or should we extend the original `ExecutorGetter` interface with:
> 
> ```go
> type ExecutorGetter interface {
> 	Get(host string) (e executor.Executor)
> 	GetSSHKeySet() (string, string)
> }
> ```

It's OK at this time. In the future there may be some executors that don't relay on ssh, at that time we can refactor this.

**Comment by lucklove** [comment]
What's the status of this? It seems still using ctx.PublicKeyPath? https://github.com/pingcap/tiup/pull/910/files#diff-c2cb8d42eb23179c5056c3d5c626c7f2ddabad3db5171c8f869d59f75ddffc42R76

**Comment by 9547** [comment]
@lucklove PTAL

**Review by lucklove — COMMENTED** [review]
Rest LGTM

**Review by lucklove — APPROVED** [review]

**Review by AstroProfundis — APPROVED** [review]

**Pull request merged**